functions/go/set-image/transformer/images_transformer.go

```diff
@@ -89,68 +98,147 @@ func (t *SetImage) configDefaultData() error {
+// matchImage takes the resources image name and container name, return if there is a match and potential warning
+func matchImage(oldImageName string, oldContainer string, newImage *Image) (bool, *fn.Result) {
+	// name would be deprecated, means image name
+	if newImage.Name != "" {
+		newImage.ImageName = newImage.Name
+	}
 
+	// match image name, no container name
+	if newImage.ImageName != "" && newImage.ContainerName == "" {
+		if !image.IsImageMatched(oldImageName, newImage.ImageName) {
+			return false, nil
+		}
+	}
+	// match container name, no image name
+	if newImage.ImageName == "" && newImage.ContainerName != "" {
+		if oldContainer != newImage.ContainerName {
+			return false, nil
+		}
+	}
+	// match both
+	if newImage.ImageName != "" && newImage.ContainerName != "" {
+		if !image.IsImageMatched(oldImageName, newImage.ImageName) {
+			return false, nil
+		}
+		//if only image name match, container name does not match, provide a warning
+		if oldContainer != newImage.ContainerName {
+			msg := fmt.Sprintf("container name `%v` does not match `%v`, only image name matches", newImage.ContainerName, oldContainer)
+			warning := &fn.Result{
+				Message:  msg,
+				Severity: fn.Warning,
+			}
+			return true, warning
+		}
+	}
+	return true, nil
+}
+
+// isImageNameUnique checks if there is only one image name in all resources, return true if name is unique
+func (t *SetImage) isImageNameUnique(items fn.KubeObjects) bool {
+	curImageName := ""
+	for _, o := range items {
+		containers := getContainers(o)
+		for _, c := range containers {
+			oldValue := c.NestedStringOrDie("image")
+			imageName, _, _ := image.Split(oldValue)
+			if curImageName == "" {
+				curImageName = imageName
+			} else if curImageName != imageName {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// updateContainerImages updates the images inside containers, return warnings and error
+func (t *SetImage) updateContainerImages(containers fn.SliceSubObjects) (fn.Results, error) {
+	var warningList fn.Results
 	for _, o := range containers {
 		oldValue := o.NestedStringOrDie("image")
-		if !image.IsImageMatched(oldValue, t.Image.Name) {
+		oldContainer := o.NestedStringOrDie("name")
+
+		matched, warning := matchImage(oldValue, oldContainer, &t.Image)
+		if !matched {
 			continue
 		}
+
 		newName := getNewImageName(oldValue, t.Image)
 		if oldValue == newName {
 			continue
 		}
 
+		if warning != nil {
+			warningList = append(warningList, warning)
+		}
+
 		if err := o.SetNestedString(newName, "image"); err != nil {
-			return err
+			return warningList, err
 		}
 		t.resultCount += 1
 	}
+	return warningList, nil
+}
+
+// getContainers gets the containers inside kubeObject
+func getContainers(o *fn.KubeObject) fn.SliceSubObjects {
+	switch o.GetKind() {
+	case "Pod":
+		return getPodContainers(o)
+	case "Deployment", "StatefulSet", "ReplicaSet", "DaemonSet", "PodTemplate":
+		return getPodSpecContainers(o)
+	}
 	return nil
 }
```

Comment:
suggest adding it back to the main `Run` method. This part belongs to the main logic so it's better to be replaced in the main function

Reply:
`Run` method already uses this function `getContianers`, I have this as a separate function because I feel I need to iterate them twice, once for validating user input, once for update the images. To reduce code duplication, I have this function here.